### ISSUE TITLE
CT Assertion failed: there should be at least as many PDOMDisplays as Displays

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/greenhouse-effect#46]
```
greenhouse-effect : phet-io-state-fuzz : unbuilt
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/phet-io-wrappers/state/?sim=greenhouse-effect&phetioDebug&fuzz&wrapperContinuousTest=%7B%22test%22%3A%5B%22greenhouse-effect%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1623792299218%22%2C%22timestamp%22%3A1623795556960%7D
Uncaught Error: Assertion failed
Error: Assertion failed
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/assert/js/assert.js:25:13)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:284:15)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removeAllPDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:236:10)
at PDOMDisplaysInfo.onVisibilityChange (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:119:14)
at LayerModelScreenView.onVisiblePropertyChange (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/scenery/js/nodes/Node.js:3736:28)
at TinyForwardingProperty.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1623792299218/axon/js/TinyEmitter.js:68:48)
id: Bayes Chrome
Snapshot from 6/15/2021, 3:24:59 PM
```

**Comment by jbphet** [comment]
This looks PDOM-related.  I'm assigning to @jessegreenberg for now, and he and I will likely take a look in our meeting today.

**Comment by KatieWoe** [comment]
This may be related:
```
greenhouse-effect : phet-io-state-fuzz : unbuilt
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/phet-io-wrappers/state/?sim=greenhouse-effect&phetioDebug&fuzz&wrapperContinuousTest=%7B%22test%22%3A%5B%22greenhouse-effect%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1624462178393%22%2C%22timestamp%22%3A1624464085611%7D
Uncaught Error: Assertion failed: there should be at least as many PDOMDisplays as Displays
Error: Assertion failed: there should be at least as many PDOMDisplays as Displays
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/assert/js/assert.js:25:13)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:284:15)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removePDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:303:35)
at PDOMDisplaysInfo.removeAllPDOMDisplays (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:236:10)
at PDOMDisplaysInfo.onVisibilityChange (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/accessibility/pdom/PDOMDisplaysInfo.js:119:14)
at WavesScreenView.onVisiblePropertyChange (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/scenery/js/nodes/Node.js:3736:28)
at TinyForwardingProperty.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1624462178393/axon/js/TinyEmitter.js:68:48)
id: Bayes Chrome
Snapshot from 6/23/2021, 9:29:38 AM
```

**Comment by jessegreenberg** [comment]
By placing a debugger in PDOMDisplaysInfo, it looks like this happens when the WavesScreenView becomes invisible (switching to the PhotonsScreenView).

**Comment by jessegreenberg** [comment]
I worked on this for a while but never got to the cause. Some notes though:
- I re-learned that PDOMDisplaysInfo is a class on each Node that tracks whether that Node has some content in the PDOM displayed under a Display which has accessibility enabled. It updates whenever the Node is added/removed, has PDOM content set/removed, visibility changes, and whenever it is added/removed from a new Display. These updates happen synchronously. It is used by PDOMInstance to control whether the actual PDOM element should be `hidden`.
- The assertion in this issue always happens when calling `removeAllPDOMDisplays` (from recursion) on the "Start Sunlight" button in the observation window.

I am not sure why this wouldn't be working in the downstream sim and will probably request some assistance from the PhET-iO team.

**Comment by jessegreenberg** [comment]
We have set `supportsInteractiveDescription` false in package.json to prevent this assertion for now to move forward with PhET-iO instrumentation.

**Comment by jessegreenberg** [comment]
Coming back to this with @zepumph, the comment to produce the in https://github.com/phetsims/greenhouse-effect/issues/46#issuecomment-870116181 is misleading, as we understand it now we still need to fuzz the sim to see this issue.

**Comment by zepumph** [comment]
We were able to reproduce this problem by checking out these shas, and then fuzzing the state wrapper.

https://phet-dev.colorado.edu/html/greenhouse-effect/1.0.0-dev.12/phet/dependencies.json

**Comment by jbphet** [comment]
Okay, here are the steps that I took in order to re-enable the a11y view without causing CT failures due to the issue with PCOMDisplays:

1. I added `"supportsInteractiveDescription": true` back to `package.json`
2. I added this sim to the phet-io-state-unsupported file
3. I added this sim back to the interactive-description file

@zepumph and @jessegreenberg - Are those the correct steps?

**Comment by zepumph** [comment]
4. Run `grunt update`. To make sure the package.json is correct in the development html.

>I added this sim back to the `interactive-description` file

While what you have listened will work. . . .
This will be done automatically each night, or via the manually run `generate-data` perennial script. Ideally this file is not edited manually.

**Comment by jbphet** [comment]
I ran `grunt update` and it didn't change `package.json`, so I think we're probably good.

I'm going to leave this open and deferred until https://github.com/phetsims/scenery/issues/1312 is taken care of, and then will follow up with turning state testing back on.

**Comment by zepumph** [comment]
A fix has been pushed over in [phetsims/scenery#1312 (comment)](https://github.com/phetsims/scenery/issues/1312#issuecomment-1047142324), next steps for this issue are to make sure it is being tested on CT in the state wrapper. This will ensure we got the fix correct.

While testing for the fix, I noticed some other state issues, so please test fuzzing the state wrapper locally also.

**Comment by jbphet** [comment]
@zepumph - I need a bit of clarification to help me figure out what the next steps are on this issue.  One thing you said was:

>  [The] next steps for this issue are to make sure it is being tested on CT in the state wrapper.

Do I need to turn on this testing?  If so, how is that done?  I just looked around in the code and found the file `phet-io-state-unsupported` that looks like it contains a list of sims for which state is unsupported, and it looks like this feeds into CT.  But greenhouse-effect isn't on this list, so I'm not sure what needs to be done for this sim.

Also, you said:

> I noticed some other state issues, so please test fuzzing the state wrapper locally also.

I just tried running state fuzz testing locally on my machine and, indeed, it hit an assertion fairly quickly.  Is this the sort of thing you're referring to?  Do I need to correct all such errors and then turn phet-io state fuzz testing on?

**Comment by zepumph** [comment]
>But greenhouse-effect isn't on this list

I see it here: https://github.com/phetsims/perennial/blob/d972159f0e77099bebb3b4ee1ab449fe422d02cb/data/phet-io-state-unsupported#L13

So what I was noting in this issue, is that although the original assertion is in all likelihood fixed, the state wrapper has other bugs in it that prevent me from knowing 100% that this original bug report is solved. Until greenhouse is removed from phet-io-state-unsupported and tested on CT, and the CT is clear of all bugs (including unrelated ones), I would be hesitant to close this issue, since it isn't my repo and isn't my issue.

There is no rush on supporting the state wrapper, but if it is possible to be supporting state while implementing other features, it may save time in the long run. 

Does that make sense?

**Comment by jbphet** [comment]
I took a little time today to see if I could resolve the errors that this sim is encountering when running the state wrapper test.  Unfortunately, the first problem it hit in each of my test runs was a general problem described in https://github.com/phetsims/phet-io/issues/1858, so I think I'm out of luck for working on this at the moment.

I'm going to defer this issue for a while, and will take another look in a couple of weeks.

**Comment by jessegreenberg** [comment]
Sorry about that. We expect https://github.com/phetsims/phet-io/issues/1858 to be fixed now.

**Comment by jbphet** [comment]
After the fix the @jessegreenberg described above (thanks!) I tested and found that errors were being encountered in the "Options" dialog.  I fixed these by opting a lot of the controls in this dialog out of registration, since it's temporary anyway.  After this I was able to run fuzz testing on the state wrapper for 10 minutes without any issues, so I have removed the sim from the list of those that don't support phet-io state.  I *believe* that this turns on state testing for it.

I'll leave this issue open for a while and will close it if it succeeds in the next several rounds of CT.
